pkg/volume/flexvolume/flexvolume_util.go

```diff
@@ -0,0 +1,142 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package flexvolume
+
+import (
+	"encoding/json"
+	"fmt"
+
+	"github.com/golang/glog"
+
+	"k8s.io/kubernetes/pkg/util/exec"
+)
+
+const (
+	initCmd     = "init"
+	setUpCmd    = "setup"
+	tearDownCmd = "teardown"
+)
+
+const (
+	// StatusSuccess represents the successful completion of operation in
+	// the backend.
+	StatusSuccess = "Success"
+	// StatusFailed represents that the operation is failed in the backend.
+	StatusFailure = "Failed"
+)
+
+// FlexVolumeDriverStatus represents the return value of the driver callout.
+type FlexVolumeDriverStatus struct {
+	// Status of the callout. One of "Success" or "Failure".
+	Status string
+	// Message is the reason for failure.
+	Message string
+	/* Device assigned by the driver. */
+	Device string `json:"device"`
+}
+
+// flexVolumeUtil is the utility structure to setup and teardown devices from
+// the host.
+type flexVolumeUtil struct{}
+
+func (u *flexVolumeUtil) init(plugin *flexVolumePlugin) error {
+	// call the init script
+	output, err := exec.New().Command(plugin.getExecutable(), initCmd).CombinedOutput()
+	if err != nil {
+		glog.Errorf("Failed to init driver: %s, error: %s", plugin.driverName, err.Error())
+		return err
+	}
+	var status FlexVolumeDriverStatus
+	if err := json.Unmarshal(output, &status); err != nil {
+		glog.Errorf("Failed to unmarshal output from driver: %s, output: %s, error: %s", plugin.driverName, output, err.Error())
+		return err
+	} else if status.Status != StatusSuccess {
+		errMsg := fmt.Sprintf("Failed to init driver: %s, status: %s, error: %s", plugin.driverName, status.Status, status.Message)
+		glog.Errorf(errMsg)
+		return fmt.Errorf("%s", errMsg)
+	}
+
+	glog.V(5).Infof("Successfully initialized driver %s", plugin.driverName)
+	return nil
+}
+
+// Attach exposes a volume on the host.
+func (u *flexVolumeUtil) setUp(f *flexVolumeBuilder, dir string) (string, error) {
+	execPath := f.execPath
+
+	var options string
+	if f.options != nil {
+		out, err := json.Marshal(f.options)
+		if err != nil {
+			glog.Errorf("Failed to marshal plugin options, error: %s", err.Error())
+			return "", err
+		}
+		if len(out) != 0 {
+			options = string(out)
+		} else {
+			options = ""
+		}
+	}
+
+	// Executable provider commamd.
+	cmd := f.runner.Command(execPath, setUpCmd, f.podNamespace, f.volName, f.podName, string(f.podUID), options)
+	output, err := cmd.CombinedOutput()
+	if err != nil {
+		glog.Errorf("Failed to setup the volume %s, output: %s, error: %s", f.volName, output, err.Error())
+		return "", err
```

Comment:
Do we expect the scripts to return a zero exit status even in error cases? I think the a JSON response provides nice additional information, but the exit status should be the primary success/fail signal.